Docs/visual testing with storybook updates

aaronsturm opened the pull request:
## Description

I was trying to configure Gastby and Storybook to work together for the first time last week. After reviewing the documentation on the Gastby site, I couldn't figure out why my fresh install looked so different. After digging into it, I noticed on January 11th the Storybook team released version 5.3.0 which had some [major changes regarding their configuration files](https://medium.com/storybookjs/declarative-storybook-configuration-49912f77b78). Although they do say it's backwards compatible, it can be confusing and I'm hoping to spare someone in the future from having to spend the same amount of time I just did researching the differences.

### Documentation

@gatsbyjs/learning this updates the existing documentation at https://www.gatsbyjs.org/docs/visual-testing-with-storybook/. I hope it meets standards as I don't consider myself a documentation expert in the slightest. I'm open and happy to make any changes to meet quality standards. I tried my best to review all of the contributing documentation before opening this pull request.

## Related Issues
At the time I started this, I didn't see any existing pull requests or issues on the changes for Storybook version 5.3.0.

Thank you!

laurieontech commented:
Thanks so much for the PR! This will no doubt help a lot of people. We'll want someone to walk through the steps to make sure they're reproducible, but should be able to complete a review on this shortly.

jonniebigodes commented:
@aaronsturm  this process as already begun with #21169 and @chrish-d ground work.

Also a side note about this, if this will be used in favour of said pr, a quick note about it. It would be a nice touch to add in a little more context to this, possibly by adding a link to the storybook documentation page, more specifically  [this](https://storybook.js.org/docs/configurations/custom-webpack-config/) to inform the reader of what is happening.

chrish-d commented:
@jonniebigodes I need to use Storybook on an upcoming project. No rush, but when the updates docs are ready I can have a run through to test and give a second opinion.

laurieontech commented:
Thanks for pointing that out @jonniebigodes. 
It does appear that issue #20624 encompasses this PR. That being said, it doesn't seem like @cristianaortiga is actively working on this.
I'll wait to see if they chime in, but I'd be inclined to have everyone contribute to the draft developed by this PR.

aaronsturm commented:
Hi @jonniebigodes and all - Thank you all for your input. I did see issue #21169 but it was specifically speaking on changing the default install to `npx` and not about the configuration changes that came about as of Storybook 5.3.0, which is the main point of this PR. That being said, I did make the `npx` install a separate commit so that it can be removed if it stomps on the word @chrish-d is doing, which of course isn't my intent.

To clarify, I did test these changes on a fresh install of Gastby using the default `header` component in Storybook. That being said, it would definitely help having a second pair of eyes to see if it makes sense to anyone else other than myself, and to make sure I didn't accidentally miss anything. :) Thanks!

jonniebigodes commented:
@aaronsturm i'm aware of the ground work being done in the pr i've mentioned above and looks like @chrish-d is also on board with this. So probably the safe bet her would be to merge the pr that @chrish-d made. You pick up on that, update your pr per the changes he's introducing and we procceed from there. 
All of you ok with this?

Also i glanced at the content real quick and so far only a couple of key points are necessary to address. As soon as i have a full blown reproduction based on this ready i'll come back and provide feedback, sounds good to all of you?

laurieontech commented:
The reason the `npx` PR hasn't been merged is because our docs try to unify on `npm`. If we need to deviate in this case it would be good to explicitly say why. I'll note that in the other PR as well.

aaronsturm commented:
Now that the other PR is merged, do you want me to rebase this branch on master?

laurieontech commented:
That'd be great. I can do it but since you have context on your changes I'd love for you to incorporate the other PR as makes the most sense.

cdvillard commented:
Piggy-backing off of this after being redirected from #21543 by @laurieontech. Would you be opposed to updating the wording of the comments? They provide context but don't read well grammatically. @wardpeet provide better suggestions than I could for these. https://github.com/gatsbyjs/gatsby/pull/21543/files/72302c0e91cc565bf25979fcda5c71a56c1337af

aaronsturm commented:
Hi @laurieontech - I've rebased this branch against the current master and removed my `npx` commit. Let me know if there's anything else I should update/fix.

@cdvillard - It might be worth re-opening your PR after this one is merged? I can fix the code comments if it's preferred.

jonniebigodes commented on docs/docs/visual-testing-with-storybook.md:
```diff
@@ -21,6 +21,130 @@ npx -p @storybook/cli sb init
+module.exports = {
+  stories: ["../stories/**/*.stories.js"],
+  addons: ["@storybook/addon-actions", "@storybook/addon-links"],
+}
```
@aaronsturm it seems that Storybook introduced a preset when you initialize it. To allow the reader to have the most accurate information, then this code block should be the following:
```javascript
module.exports = {
  stories: ["../stories/**/*.stories.js"],
  addons: ['@storybook/preset-create-react-app','@storybook/addon-actions', '@storybook/addon-links'],
};
```

aaronsturm replied:
@jonniebigodes - Yes block was the default `.storybook/main.js` when I ran `npx -p @storybook/cli sb init`. I have 3 sections, one is what the user would start with, one is with the changes to introduce, and one is the final product all together. I did this to help bring clarity.

I haven't needed to add `'@storybook/preset-create-react-app'` to make this work. Is it required?

jonniebigodes replied:
the preset is added by default when storybook is initialized as per a recent release. I mentioned it to keep the documentation as accurate as much as possible.

jonniebigodes commented on docs/docs/visual-testing-with-storybook.md:
```diff
@@ -21,6 +21,130 @@ npx -p @storybook/cli sb init
+import { action } from "@storybook/addon-actions"
+
+// automatically import all files ending in *.stories.js
+// configure(require.context("../src", true, /\.stories\.js$/), module)
```
@aaronsturm if you can remove these two lines as they are not accurate and no longer needed.

aaronsturm replied:
Thanks for catching that, I had thought I removed those lines :(.

jonniebigodes commented on docs/docs/visual-testing-with-storybook.md:
````diff
@@ -21,6 +21,130 @@ npx -p @storybook/cli sb init
+config.resolve.extensions.push(".ts", ".tsx")
+```
+
+## Storybook versions 5.0, 5.1, and 5.2
````
@aaronsturm if you can remove this portion it would be good. On a recent conversation with one of the Storybook maintainers, he brought to my attention that they're pushing for version >=5.3 as much as possible, to provide early adoption as much as possible. With that in mind i coordinated with him to update some of the tutorials to address this push. Now when a reader  sees this document and follows it, he/she will be using the latest version with all it's features. If by any chance the reader decides to go and lookup the starter library and doing a quick check on them, it seems that only one is currently updated. more specifically [this](https://github.com/georgewilde/gatsby-starter-wilde-creations), the other ones available haven't been updated from 3 months to almost a year. So with that in mind the reader should be prompted to direct any issues to the respective repos and bring to the attention of their maintainers that he has some issues running any of the starters and with that probably avoid creating issues in the Gatsby repo that could be not related. Sounds good? Also @laurieontech are you on board with this?

jonniebigodes replied:
Regarding 4.0 section. It should be removed as there's alot of breaking changes between 4.x and 5.2 and from 5.2 to 5.3. When the reader follows the documentation he/she will be using the latest release of the package.

Regarding 5.2 i would leave it in but with a word of advice regarding the starters as most of them are not currently updated. Like i mentioned above.

aaronsturm commented:
@cdvillard - I just pushed changes that has comment updates from your PR, with the recommendations from @wardpeet.

 @jonniebigodes - I made the changes you request (I hope I got them all!) and I also added a missing package that's needed for TypeScript. I re-ran through the steps in a clean folder to verify these steps.

Let me know if there's anything else. Thanks!

aaronsturm commented:
Hi @jonniebigodes & @laurieontech - I wanted to follow-up on this to see if you saw my updates a few days ago? Thanks!

jonniebigodes commented:
@aaronsturm sorry but i was sidetracked, if you don't mind i'll take a closer look tomorrow if you don't mind.

laurieontech commented:
Heads up that this is on our list to review. We'll do a full test of the instructions as well.

cdvillard commented:
@laurieontech, I tried these instructions out with a new project I'm working on and, and it seems to work, at least prior to proper testing. Tangentially, another issue may need to be addressed as far as components imported from the Gatsby project using `useStaticQuery.` I found that mine would error out.

laurieontech commented:
@cdvillard That's great context. Mind opening a bug report with a reproduction?

jonniebigodes commented:
I've had the chance to go through this and the only thing missing is update the "Writing stories section" of the documentation, the current example is using the old format that is still available for the moment but the team is pushing for what they call CSF (Component Story Format), you can read more about it in [here](https://storybook.js.org/docs/formats/component-story-format/) and [here](https://medium.com/storybookjs/component-story-format-66f4c32366df). If you can make the change it would be great.

@cdvillard i wouldn't mind taking a look at it. But see if it relates to #20624 it could be that the data isn't materialized a priori and it breaks Storybook.

marcysutton commented on docs/docs/visual-testing-with-storybook.md:
```diff
@@ -21,97 +21,200 @@ npx -p @storybook/cli sb init
 
-  // Prefer Gatsby ES6 entrypoint (module) over commonjs (main) entrypoint
-  config.resolve.mainFields = ["browser", "module", "main"]
+After line 25 containing `config.resolve.mainFields = ["browser", "module", "main"];` and before line 27 containing `return config;`
```
To make sure we don't introduce docs debt or confusion: is there a more general way to describe the job to be done in case the user has existing info in their Storybook configuration (i.e. can you count on line numbers here)?

aaronsturm replied:
Hi @marcysutton - My thought process is the user could compare this to a fresh install of `main.js` to figure out where to put the lines if their configuration differed. I did this because when I went through this initially, I kept thinking, "Where am I supposed to put this?!" Using line numbers was the only way I could think of to accurately explain where to put the new configuration, but I'm open to changing this if you feel it will cause issues going forward.

marcysutton replied:
That makes sense–I'm mostly wondering if counting on line numbers from a fresh install would be brittle, since that file is generated by Storybook. If you describe it more generally _and_ hint with line numbers, would it still work?

```suggestion
Add the following code after the line containing `config.resolve.mainFields = ["browser", "module", "main"];` (line 25) and before the `return config;` in the same function body (line 27):
```

jonniebigodes commented:
A couple of quick questions...why is the version 4.0 when the flow of the documentation leads towards setting up storybook with gatsby? At most point to the migration docs. 

And one key point that was referenced in the issue that kickstarted this whole endeavour is not being mentioned whatsoever. More specifically that the reader needs to generate a build when using the graphql data layer provided by gatsby or he/she will run into problems?

jonniebigodes commented on docs/docs/visual-testing-with-storybook.md:
````diff
@@ -21,97 +21,200 @@ npx -p @storybook/cli sb init
+}
+```
+
+Adjustments to Storybook's default `webpack` configuration are required so that you can transpile Gatsby source files and to ensure you have the necessary `babel` plugins to transpile Gatsby components. Add the following section to `.storybook/main.js`.
````
@marcysutton the preset was added due to the fact that i bumped the version to 6.0.x alpha to check if all would work in the same way as it's working with the current release

aaronsturm replied:
@marcysutton - Updated the wording.

PaulieScanlon commented on docs/docs/visual-testing-with-storybook.md:
```diff
@@ -21,97 +21,200 @@ npx -p @storybook/cli sb init
+      require.resolve("@babel/preset-env"),
+    ]
+
+    config.module.rules[0].use[0].options.plugins = [
```
I'm experiencing issues with the Storybook prop table not populating. 
Commenting out this line allows the PropTable to work but brings back GraphQL errors... any ideas?

PaulieScanlon replied:
i've found that if i manually add `babel-plugin-react-docgen` the plugins the PropTable works again.. perhaps these options need to be spread before / after adding `@babel/plugin-proposal-class-properties` and `babel-plugin-remove-graphql-queries`


This is my final config:


```javascript
  webpackFinal: async config => {
    // Transpile Gatsby module because Gatsby includes un-transpiled ES6 code.
    config.module.rules[0].exclude = [/node_modules\/(?!(gatsby)\/)/];

    // use installed babel-loader which is v8.0-beta (which is meant to work with @babel/core@7)
    config.module.rules[0].use[0].loader = require.resolve("babel-loader");

    // use @babel/preset-react for JSX and env (instead of staged presets)
    config.module.rules[0].use[0].options.presets = [
      require.resolve("@babel/preset-react"),
      require.resolve("@babel/preset-env"),
    ];

    config.module.rules[0].use[0].options.plugins = [
      // use @babel/plugin-proposal-class-properties for class arrow functions
      require.resolve("@babel/plugin-proposal-class-properties"),

      // use babel-plugin-remove-graphql-queries to remove static queries from components when rendering in storybook
      require.resolve("babel-plugin-remove-graphql-queries"),

      // use babel-plugin-react-docgen to ensure PropTables still appear
      require.resolve("babel-plugin-react-docgen"),
    ];

    // Prefer Gatsby ES6 entrypoint (module) over commonjs (main) entrypoint
    config.resolve.mainFields = ["browser", "module", "main"];

    return config;
  },

```

jonniebigodes replied:
@PaulieScanlon have you tried issuing a build, a development build for instance before running Storybook? More on that in #20624.

PaulieScanlon replied:
Yeah i did run in to issues if i'd recently cleaned Gatsby so running a dev first and then Storybook with the doc-gen back in config seems to work but i have noticed it blats out all other webpack config so is there a way to spread the plugins before applying the babel plugins so it maintains some of the original config?

marcysutton replied:
@PaulieScanlon If that issue is still persisting with these changes applied, can you open a new issue to track it? Thanks!

marcysutton commented:
To answer @jonniebigodes' questions:

> why is the version 4.0 when the flow of the documentation leads towards setting up storybook with gatsby? At most point to the migration docs.

A good question that I don't yet have a good answer to: is anyone still using Storybook version 4 for any reason? It's not a huge lift to keep around, but a worthwhile question. One of my PR review comments was about introducing the Version 4 instructions, so perhaps that's a good place to discuss why someone might need to know about Version 4 at all if it is still valuable to keep.

> And one key point that was referenced in the issue that kickstarted this whole endeavour is not being mentioned whatsoever. More specifically that the reader needs to generate a build when using the graphql data layer provided by gatsby or he/she will run into problems?

I don't see what you're referring to in this PR description, is there a separate issue you have in mind (we recommend opening issues first, in general)? It would certainly make sense to include information on any gotchas like needing to build first, so that makes sense to add.

jonniebigodes commented:
@marcysutton to give you some more context.

Regarding this:

> I don't see what you're referring to in this PR description, is there a separate issue you have in mind (we recommend opening issues first, in general)? It would certainly make sense to include information on any gotchas like needing to build first.

If you followed the trail of comments and links that were posted you would end up with #20624.  I answered there and told the author of said issue to make the change and contribute to the improvement of the said documentation covered in this pr and also mentioned the problem.

To sum it up the problem extends not only to usage of the data layer, but also for a simple Gatsby site that doesn't use it at all. When the reader follows the documentation and adds Storybook to a Gatsby site which is based in React when he/she starts Storybook without having a previous build generated with either `gatsby develop` or `gatsby build` Storybook will break as it's expecting the `public` folder to exist and it's contents materialized. And that tiny bit of crucial information is not present in the documentation whatsoever, hence why i mentioned so that it would prevent more issues to be created in this repo, that could be easily solved by the including that piece of information.

aaronsturm commented on docs/docs/visual-testing-with-storybook.md:
````diff
@@ -21,97 +21,200 @@ npx -p @storybook/cli sb init
+}
+```
+
+Next create a new file under `.storybook` called `preview.js`. This configuration file preview.js is not responsible for loading any stories. Its main purpose is to add global parameters and decorators.
````
@marcysutton - I actually stole that wording from Storybook themselves under the heading preview.js at https://medium.com/storybookjs/declarative-storybook-configuration-49912f77b78

I made decorators a link to [https://storybook.js.org/docs/addons/introduction/#1-decorators](https://storybook.js.org/docs/addons/introduction/#1-decorators). Do you feel that will resolve any confusion for people new to Storybook?

marcysutton replied:
The link helps, for sure!

aaronsturm commented on docs/docs/visual-testing-with-storybook.md:
```diff
@@ -21,97 +21,200 @@ npx -p @storybook/cli sb init
 This command adds a set of boilerplate files for Storybook in your project. However, since this is for a Gatsby project, you need to update the default Storybook configuration a bit so you don't get errors when trying to use Gatsby specific components inside of the stories.
 
-To update your Storybook config open `.storybook/config.js` and modify the content as follows:
+## Storybook version 5
```
Done

aaronsturm commented on docs/docs/visual-testing-with-storybook.md:
````diff
@@ -21,97 +21,200 @@ npx -p @storybook/cli sb init
 ```
 
-**For Storybook v4:**
+## Storybook version 4
````
Done

aaronsturm commented on docs/docs/visual-testing-with-storybook.md:
```diff
@@ -21,97 +21,200 @@ npx -p @storybook/cli sb init
-> You can remove the `stories` folder from the root of your project, or move it inside your `src` folder
-
-Next make some adjustments to Storybook's default `webpack` configuration so you can transpile Gatsby source files, and to ensure you have the necessary `babel` plugins to transpile Gatsby components.
+### Add TypeScript Support
```
Done

aaronsturm commented on docs/docs/visual-testing-with-storybook.md:
```diff
@@ -21,97 +21,200 @@ npx -p @storybook/cli sb init
 // Gatsby's Link overrides:
-// Gatsby defines a global called ___loader to prevent its method calls from creating console errors you override it here
+// Gatsby Link calls the `enqueue` & `hovering` methods on the global variable ___loader.
+// This global object isn't set in storybook context. We override it to empty functions (no-op),
```
My bad, that was copy and paste from [https://github.com/gatsbyjs/gatsby/pull/21543](https://github.com/gatsbyjs/gatsby/pull/21543). It has been updated.

aaronsturm commented:
>  I've had the chance to go through this and the thing missing is update the "Writing stories section" of the documentation, the current example is using the old format that is still available for the moment but the team is pushing for what they call CSF (Component Story Format), you can read more about it in here and here. If you can make the change it would be great.

@jonniebigodes - I took the example that was there from the previous author and converted it to CSF. I believe I have captured all of your recommendations at this point?

> Is anyone still using Storybook version 4 for any reason? 

@marcysutton - I don't think so as it's a 2 year old version of Storybook. Since it was already written by the previous author, I figure it didn't hurt to leave it there in a "just in case" scenario. Maybe when 6.0 comes out it can be removed? If the Gatsby docs want to only keep one previous Storybook version?

I believe I have updated this PR with everyone's feedback, although it's possible I missed something while scrolling. Let me know if I missed anything. Thanks!

marcysutton commented on docs/docs/visual-testing-with-storybook.md:
````diff
@@ -21,97 +21,200 @@ npx -p @storybook/cli sb init
 ```
 
-**For Storybook v4:**
+### Storybook version 4
````
```suggestion
With setup completed for Storybook 5, you can continue with the [information below to write stories](#writing-stories).

### Storybook version 4

To use Storybook version 4 with Gatsby, use the following setup instructions:
```

This section still jumps straight from a heading into source code, so even if it's getting removed later we should make it read better in the meantime.

marcysutton commented on docs/docs/visual-testing-with-storybook.md:
```diff
@@ -170,13 +273,16 @@ First, create the story file. Storybook looks for all files with a `.stories.js`
 import React from "react"
-import { storiesOf } from "@storybook/react"
 
-storiesOf(`Dashboard/Header`, module).add(`default`, () => (
```
I see the `storiesOf` API isn't deprecated yet, but will this code work in a backwards compatible fashion? https://storybook.js.org/docs/formats/storiesof-api/

jonniebigodes replied:
@marcysutton the `storiesOf` api will still be around for a bit as some other frameworks haven't still been ported to use CSF. More specifically the react native is still using it to declare the stories format and with that for Storybook to consume them. And CSF was introduced in version 5.2. Prior to that CSF approach will not work. Having both formats seems counter productive at this day and age. More even because the documentation is pointing to CSF usage.

marcysutton commented:
> Maybe when 6.0 comes out it can be removed? If the Gatsby docs want to only keep one previous Storybook version?

That sounds good to me. 

To address @jonniebigodes's points about running `gatsby build` before Storybook can do its thing, can you add something about that to the "Writing stories" section? That seems like a straightforward change that could prevent some gotchas.

Otherwise, I left a one suggestion and one question about the test example but this is really close. Thanks for your patience on it as we've been undergoing some team changes. I'd love to get this merged ASAP!

PaulieScanlon commented:
Not sure If this issue is still active or if this is related but there are some issues with this setup of you’re running it in Yarn workspaces. Perhaps the docs should mention this? Here’s my current investigation. 

https://github.com/PaulieScanlon/typescript-theme-ui-link/issues/1

marcysutton commented:
@PaulieScanlon the Gatsby docs officially recommend npm at this time, with Theme authoring as one notable exception. If there's a clear fix that could be worked in quickly, that would be awesome. But my hunch is that it should probably be opened as a separate issue to recommend a single pathway of advice if possible, without specifically calling out Yarn.

PaulieScanlon commented:
Hey @marcysutton i'm not sure if this is a `yarn` or a `npm` related issue and having got something working with `yarn workspaces` for both `js` and `tsx` leads me to belive the example webpack config in these docs are not quite right... or at least if you follow these docs you'll get TypeScript errors in a normal repo and you'l get JavaScript and TypeScript errors if you run this in a workspace... This [repo](https://github.com/PaulieScanlon/typescript-theme-ui-link/blob/master/demo/.storybook/main.js) appears to solve it for both cases so if one of your guys has time to look at it and confirm i've not made a gross mess of the webpack config maybe this doc could be updated to show my webpack config?

aaronsturm commented:
@PaulieScanlon - The Webpack config that's in this is the same that's on the current site: [https://www.gatsbyjs.org/docs/visual-testing-with-storybook/](https://www.gatsbyjs.org/docs/visual-testing-with-storybook/). This is based on a fresh `create-react-app` install, just like the [official Storybook documentation](https://storybook.js.org/docs/configurations/typescript-config/). It would be impossible to know everyone's custom Webpack configurations and how to fix their specific issue. This PR is specifically about updating the documentation to reflect the [Storybook 5.3 configuration file name differences](https://storybook.js.org/docs/configurations/typescript-config/).

PaulieScanlon commented:
@aaronsturm Hey man! Yeah my `js` webpack is pretty much the same as the current docs but i also added a TypeScript example.. and more crucially how both can co-exist. You are right though it's impossible to know how every config would work but for me having something clear for TypeScript in the Gatsby docs would be helpful. The only real difference between the two in my example are the doc-gen loaders... oh and the prop inheritance filter. No worries if this isn't of help... I just know i've had this problem a few times with TypeScript | Storybook and  Gatsby and am glad i've finally solved it

aaronsturm commented:
@PaulieScanlon - A section on how to add TypeScript support does exist, are you saying you did that and still had issues?. I have gone through this at least 10 times installing a fresh copy and going through this step by step. It is possible that through familiarity I automatically assumed a step, or some other human error, so do let me know if there's anything that's missing.

> The only real difference between the two configs in my example are the "doc-gen" loaders...

I didn't add doc-gen as that's a separate install from the base `create-react-app`.

PaulieScanlon commented:
@aaronsturm just a quick slightly un-related / related question... what has `create-react-app` got to do with this?

jonniebigodes commented:
@aaronsturm i'm onboard with you. The proposed documentation change is meant to cover the general usage of both frameworks and i'm inclined to leave it as is. @PaulieScanlon yours is a more subtle one. Usage of monorepos with either yarn or npm should be something that would probably best suited for the storybook documentation

PaulieScanlon commented:
@jonniebigodes that's a fair point. I wonder then if it's worth noting in this doc that this config only applies if it's in a standard node project (excluding workspaces) and if you attempt to add TypeScript support you may also run in to problems with this setup. Which was basically my problem.


I needed yarn workspace `.js` and `.tsx` support. My webpack config will do all of those things and will work if it's **not** in a workspace **and *if it's only `.js` or **only** `.ts`

aaronsturm commented:
> @aaronsturm just a quick slightly un-related / related question... what has `create-react-app` got to do with this?

@PaulieScanlon - Storybook was originally designed to work with React before they adopted other frameworks. I know React, so it's easiest for me to run `create-react-app` to get a fresh copy of a React project.

aaronsturm commented:
> @jonniebigodes that's a fair point.
> 
> I wonder then if it's worth noting in these docs that this config only applies if it's in a standard node project (excluding workspaces) and if you attempt to add TypeScript support you may also run in to problems with this setup. Which was basically my problem.
> 
> I needed yarn workspaces `.js` and `.tsx` support.
> 
> My example webpack config will do all of those things and will work if it's **not** in a workspace **and** if it's only `.js` or **only** `.tsx` **and** both... **and** includes "doc gen" information which is kinda helpful in TypeScript.
> 
> It's cool if this too much. Maybe i should just write a blog post about it ay 😊

@PaulieScanlon - I personally prefer to have documentation with the least amount of configuration as necessary. For instance, why would I want docgen support if I don't use it, or use [TypeDoc](https://typedoc.org/), or something else?

PaulieScanlon commented:
Yeah fair point again. Perhaps these docs could just be more specific about the scenarios in which it’ll work and what’s been left out. I think I’ve I’d know about the workspace thing to begin with I wouldn’t have been so confused. Unless this config does work in a mono repo?. I chased my tail for a while with this so may have gone round in circles.

jonniebigodes commented:
@aaronsturm once again i'm onboard with you. @PaulieScanlon having alot of code and cases makes the reader "digress" and loose track of the focus of the documentation he/she is reading. If you both agree with this, a good middle ground that could be achieved is to add a note that this document only covers a typical gatsby/storybook setup, non monorepos with yarn or npm. Sounds good?

PaulieScanlon commented:
@jonniebigodes yeah perfect, a note to say it doesn't work with mono repos ... if that is definitely the case. 

It's probably quite relevant in Gatsby world since Themes are all about workspaces and i know it would have saved me some time if i'd known that up front.

aaronsturm commented:
@jonniebigodes - I'm fine with that, but now I am curious to try it out and see why it wouldn't work as is. Gatsby Themes is not something I've gotten into yet. @marcysutton - What do you think? Is this good as-is or do you want me to add the extra description?

marcysutton commented:
@aaronsturm if it's simple enough to add a little warning there in a small amount of time, it would be nice to prevent people from spinning their wheels trying to solve an issue.

But I should also plug separate issues again to tackle follow-on work rather than holding up this PR, if that proves too difficult. In fact, I think the TypeScript configuration entirely could have been shipped as a separate issue/PR. But the ship has sailed for that. Thanks for all the discussion here!

aaronsturm commented:
@marcysutton  - I used this article to start testing: https://www.gatsbyjs.org/blog/2019-05-22-setting-up-yarn-workspaces-for-theme-development/ and right from the start, adding Storybook with its recommended `npx -p @storybook/cli sb init` caused issues due to Yarn complaining about adding Storybook to the project root with Workspaces enabled. I don't know enough about the best practices with Storybook and Gastby Themes/Yarn Workspaces to state anything official.

Based on what you said in your last comment, I would say merge this and create a separate issue/PR in the future to add a section on using Gatsby Themes and Storybook together. I do see that extra information being useful to someone using Themes. Thanks!

marcysutton commented:
Sounds good, I'd say let's leave that for a future issue then. Thanks so much for your work on this!

marcysutton commented:
The one thing that might be worth adding still though, is a note about running `gatsby build`. What do you think?

aaronsturm commented:
> The one thing that might be worth adding still though, is a note about running `gatsby build`. What do you think?

At which step? This might be one of those things I did through familiarity. But I did test every step myself multiple times.

marcysutton commented:
> At which step?

Before building tests the first time, I believe. That's what @jonniebigodes mentioned earlier...do you think that applies here? If not, we should go ahead and merge this.

aaronsturm commented:
> > At which step?
> 
> Before building tests the first time, I believe. That's what @jonniebigodes mentioned earlier...do you think that applies here? If not, we should go ahead and merge this.

Let's go ahead and merge this so that others can get more up-to-date documentation and hopefully not go through the process I did. Any additional modifications/streamlines can be faster and smaller PRs. That's my 2 cents anyway.

aaronsturm commented:
Hi @marcysutton - Is there anything else that I need to do? I'd like to get this merged in. I can quickly rebase it onto master if you would like? Thanks!
